packit/specfile.py

```diff
@@ -170,58 +170,61 @@ def add_patches(self, patch_list: List[PatchMetadata]) -> None:
-        for patch in patch_list:
-            if patch.present_in_specfile:
-                original_patches.append(patch)
-            else:
-                source_git_patches.append(patch)
-
-        logger.debug(
-            f"Original patches ({len(original_patches)}) "
-            "has to be already in the spec-file. "
-            "Following patches will not be added to the spec-file:\n - "
-            + "\n - ".join(f"{patch.name} ({patch.path})" for patch in original_patches)
-            + "\n"
-        )
+        # we could have generated patches before (via git-format-patch)
+        # so let's reload the spec
+        self.reload()
 
-        logger.debug(f"Adding source-git patches ({len(source_git_patches)})")
-        new_content = "\n# PATCHES FROM SOURCE GIT:\n"
-        patch_number_offset = len(original_patches)
-        for i, patch_metadata in enumerate(source_git_patches):
-            new_content += "\n# " + "\n# ".join(
-                patch_metadata.specfile_comment.split("\n")
-            )
-            new_content += (
-                f"\nPatch{(i + 1 + patch_number_offset):04d}: {patch_metadata.name}\n"
-            )
+        applied_patches: Dict[str, PatchObject] = {
+            p.get_patch_name(): p for p in self.get_applied_patches()
+        }
 
-        # valid=None: take any SourceX even if it's disabled
-        last_source_tag_line = [
-            t.line for t in self.tags.filter(name="Source*", valid=None)
-        ][-1]
-        # find the first empty line after last_source_tag
-        for i, line in enumerate(
-            self.spec_content.section("%package")[last_source_tag_line:]
-        ):
-            if line.strip() == "":
-                break
-        else:
-            logger.error("Can't find where to add patches.")
-            return
-        where = last_source_tag_line + i
-        # insert new content below last Source
-        self.spec_content.section("%package")[where:where] = new_content.split("\n")
+        for patch_metadata in patch_list:
+            if patch_metadata.present_in_specfile:
+                logger.debug(
+                    f"Patch {patch_metadata.name} is already present in the spec file."
+                )
+                continue
+
+            if patch_metadata.name in applied_patches:
+                logger.debug(
+                    f"Patch {patch_metadata.name} is already defined in the spec file."
+                )
+                continue
+
+            self.add_patch(patch_metadata)
 
-        logger.info(
-            f"{len(source_git_patches)}/{len(patch_list)} patches added to {self.path!r}."
+    def add_patch(self, patch_metadata: PatchMetadata):
+        """
+        Add provided patch to the spec file:
+         * Set Patch index to be +1 than the highest index of an existing specfile patch
+         * Place the patch in the spec after the last existing patch or before the %package section
```

Comment:
I'm reading the code as "after the last existing Patch or Source"

Reply:
I'll make it more clear